Compiler Internal Error: Operators should be desugared by the point of underscore to lambda conversion

### Discord username

_No response_

### What type of issue is this?

Permanent – Occurring repeatably

### Is this issue blocking you from using Enso?

- [ ] Yes, I can't use Enso because of this issue.

### Is this a regression?

- [ ] Yes, previous version of Enso did not have this issue.

### What issue are you facing?

Compiler crashes on certain syntax.

The code was something like this:
```
main =
    year = "2022"
  ma ==ums== 
```
(yes, it's invalid syntax - I blame bad syncing)

### Expected behaviour

No crash

### How we can reproduce it?

No clue tbh

### Screenshots or screencasts

_No response_

### Logs

tons of repeated logs (this isn't the entire log, it's repeated many more times)
<details>
<summary>logs</summary>
<pre><code>[WARN] [2023-12-01T13:37:59+10:00] [enso.org.enso.interpreter.service.ExecutionService] Execution of function main failed (Compiler Internal Error: Operators should be desugared by the point of underscore to lambda conversion.).
org.enso.compiler.core.CompilerError: Compiler Internal Error: Operators should be desugared by the point of underscore to lambda conversion.
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:288)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Section$Right.transformExpressions(Section.scala:235)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$desugarApplication$2(LambdaShorthandToLambda.scala:188)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:187)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Application$Prefix.transformExpressions(Application.scala:26)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$desugarApplication$2(LambdaShorthandToLambda.scala:188)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:187)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Application$Prefix.transformExpressions(Application.scala:26)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.$anonfun$transformExpressions$1(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.mapExpressions(Expression.scala:141)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.mapExpressions(Expression.scala:58)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.transformExpressions(Expression.scala:58)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.$anonfun$transformExpressions$1(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.mapExpressions(Function.scala:132)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.mapExpressions(Function.scala:46)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.transformExpressions(Function.scala:46)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.runExpression(LambdaShorthandToLambda.scala:111)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$runModule$2(LambdaShorthandToLambda.scala:81)
	at org.enso.runtime/org.enso.compiler.core.ir.module.scope.definition.Method$Explicit.mapExpressions(Method.scala:156)
	at org.enso.runtime/org.enso.compiler.core.ir.module.scope.definition.Method$Explicit.mapExpressions(Method.scala:50)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$runModule$1(LambdaShorthandToLambda.scala:79)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:250)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.runModule(LambdaShorthandToLambda.scala:75)
	at org.enso.runtime/org.enso.compiler.pass.PassManager.$anonfun$runPassesOnModule$2(PassManager.scala:101)
	at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:183)
	at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:179)
	at org.enso.runtime/scala.collection.immutable.List.foldLeft(List.scala:79)
	at org.enso.runtime/org.enso.compiler.pass.PassManager.runPassesOnModule(PassManager.scala:92)
	at org.enso.runtime/org.enso.compiler.Compiler.runGlobalTypingPasses(Compiler.scala:806)
	at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$7(Compiler.scala:309)
	at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$7$adapted(Compiler.scala:294)
	at org.enso.runtime/scala.collection.immutable.List.foreach(List.scala:333)
	at org.enso.runtime/org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:294)
	at org.enso.runtime/org.enso.compiler.Compiler.go$1(Compiler.scala:225)
	at org.enso.runtime/org.enso.compiler.Compiler.runInternal(Compiler.scala:232)
	at org.enso.runtime/org.enso.compiler.Compiler.run(Compiler.scala:127)
	at org.enso.runtime/org.enso.interpreter.runtime.Module.compile(Module.java:382)
	at org.enso.runtime/org.enso.interpreter.runtime.Module.compileScope(Module.java:309)
	at org.enso.runtime/org.enso.interpreter.service.ExecutionService.prepareFunctionCall(ExecutionService.java:119)
	at org.enso.runtime/org.enso.interpreter.service.ExecutionService.execute(ExecutionService.java:241)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.executeProgram(ProgramExecutionSupport.scala:126)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.$anonfun$runProgram$3(ProgramExecutionSupport.scala:253)
	at org.enso.runtime/scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.enso.runtime/cats.syntax.EitherObjectOps$.catchNonFatal$extension(either.scala:391)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.$anonfun$runProgram$2(ProgramExecutionSupport.scala:253)
	at org.enso.runtime/scala.util.Either.flatMap(Either.scala:352)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.runProgram(ProgramExecutionSupport.scala:247)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:43)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:18)
	at org.enso.runtime/org.enso.interpreter.instrument.execution.JobExecutionEngine.$anonfun$runInternal$1(JobExecutionEngine.scala:124)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.access$001(PolyglotThread.java:53)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$1.execute(PolyglotThread.java:106)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.executeImpl(PolyglotThread.java:140)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.execute(PolyglotThread.java:131)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:746)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:602)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:519)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:500)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.run(PolyglotThread.java:102)
[WARN] [2023-12-01T13:37:59+10:00] [enso.org.enso.interpreter.service.ExecutionService] Execution of function main failed (Compiler Internal Error: Operators should be desugared by the point of underscore to lambda conversion.).
org.enso.compiler.core.CompilerError: Compiler Internal Error: Operators should be desugared by the point of underscore to lambda conversion.
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:288)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Section$Right.transformExpressions(Section.scala:235)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$desugarApplication$2(LambdaShorthandToLambda.scala:188)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:187)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Application$Prefix.transformExpressions(Application.scala:26)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$desugarApplication$2(LambdaShorthandToLambda.scala:188)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:187)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Application$Prefix.transformExpressions(Application.scala:26)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.$anonfun$transformExpressions$1(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.mapExpressions(Expression.scala:141)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.mapExpressions(Expression.scala:58)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.transformExpressions(Expression.scala:58)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.$anonfun$transformExpressions$1(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.mapExpressions(Function.scala:132)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.mapExpressions(Function.scala:46)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.transformExpressions(Function.scala:46)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.runExpression(LambdaShorthandToLambda.scala:111)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$runModule$2(LambdaShorthandToLambda.scala:81)
	at org.enso.runtime/org.enso.compiler.core.ir.module.scope.definition.Method$Explicit.mapExpressions(Method.scala:156)
	at org.enso.runtime/org.enso.compiler.core.ir.module.scope.definition.Method$Explicit.mapExpressions(Method.scala:50)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$runModule$1(LambdaShorthandToLambda.scala:79)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:250)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.runModule(LambdaShorthandToLambda.scala:75)
	at org.enso.runtime/org.enso.compiler.pass.PassManager.$anonfun$runPassesOnModule$2(PassManager.scala:101)
	at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:183)
	at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:179)
	at org.enso.runtime/scala.collection.immutable.List.foldLeft(List.scala:79)
	at org.enso.runtime/org.enso.compiler.pass.PassManager.runPassesOnModule(PassManager.scala:92)
	at org.enso.runtime/org.enso.compiler.Compiler.runGlobalTypingPasses(Compiler.scala:806)
	at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$7(Compiler.scala:309)
	at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$7$adapted(Compiler.scala:294)
	at org.enso.runtime/scala.collection.immutable.List.foreach(List.scala:333)
	at org.enso.runtime/org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:294)
	at org.enso.runtime/org.enso.compiler.Compiler.go$1(Compiler.scala:225)
	at org.enso.runtime/org.enso.compiler.Compiler.runInternal(Compiler.scala:232)
	at org.enso.runtime/org.enso.compiler.Compiler.run(Compiler.scala:127)
	at org.enso.runtime/org.enso.interpreter.runtime.Module.compile(Module.java:382)
	at org.enso.runtime/org.enso.interpreter.runtime.Module.compileScope(Module.java:309)
	at org.enso.runtime/org.enso.interpreter.service.ExecutionService.prepareFunctionCall(ExecutionService.java:119)
	at org.enso.runtime/org.enso.interpreter.service.ExecutionService.execute(ExecutionService.java:241)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.executeProgram(ProgramExecutionSupport.scala:126)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.$anonfun$runProgram$3(ProgramExecutionSupport.scala:253)
	at org.enso.runtime/scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.enso.runtime/cats.syntax.EitherObjectOps$.catchNonFatal$extension(either.scala:391)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.$anonfun$runProgram$2(ProgramExecutionSupport.scala:253)
	at org.enso.runtime/scala.util.Either.flatMap(Either.scala:352)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ProgramExecutionSupport$.runProgram(ProgramExecutionSupport.scala:247)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:43)
	at org.enso.runtime/org.enso.interpreter.instrument.job.ExecuteJob.run(ExecuteJob.scala:18)
	at org.enso.runtime/org.enso.interpreter.instrument.execution.JobExecutionEngine.$anonfun$runInternal$1(JobExecutionEngine.scala:124)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.access$001(PolyglotThread.java:53)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$1.execute(PolyglotThread.java:106)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.executeImpl(PolyglotThread.java:140)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.execute(PolyglotThread.java:131)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:746)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:602)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:519)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:500)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.run(PolyglotThread.java:102)
[WARN] [2023-12-01T13:37:59+10:00] [enso.org.enso.interpreter.service.ExecutionService] Error while ensureCompiledModule local.AOC_2023.Main
org.enso.compiler.core.CompilerError: Compiler Internal Error: Operators should be desugared by the point of underscore to lambda conversion.
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:288)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Section$Right.transformExpressions(Section.scala:235)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$desugarApplication$2(LambdaShorthandToLambda.scala:188)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:187)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Application$Prefix.transformExpressions(Application.scala:26)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$desugarApplication$2(LambdaShorthandToLambda.scala:188)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarApplication(LambdaShorthandToLambda.scala:187)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:128)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$$anonfun$desugarExpression$1.applyOrElse(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:22)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.expression.Application$Prefix.transformExpressions(Application.scala:26)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.$anonfun$transformExpressions$1(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.mapExpressions(Expression.scala:141)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.mapExpressions(Expression.scala:58)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression$Block.transformExpressions(Expression.scala:58)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.$anonfun$transformExpressions$1(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.mapExpressions(Function.scala:132)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.mapExpressions(Function.scala:46)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions(Expression.scala:24)
	at org.enso.runtime/org.enso.compiler.core.ir.Expression.transformExpressions$(Expression.scala:18)
	at org.enso.runtime/org.enso.compiler.core.ir.Function$Lambda.transformExpressions(Function.scala:46)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.desugarExpression(LambdaShorthandToLambda.scala:126)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.runExpression(LambdaShorthandToLambda.scala:111)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$runModule$2(LambdaShorthandToLambda.scala:81)
	at org.enso.runtime/org.enso.compiler.core.ir.module.scope.definition.Method$Explicit.mapExpressions(Method.scala:156)
	at org.enso.runtime/org.enso.compiler.core.ir.module.scope.definition.Method$Explicit.mapExpressions(Method.scala:50)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.$anonfun$runModule$1(LambdaShorthandToLambda.scala:79)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:250)
	at org.enso.runtime/org.enso.compiler.pass.desugar.LambdaShorthandToLambda$.runModule(LambdaShorthandToLambda.scala:75)
	at org.enso.runtime/org.enso.compiler.pass.PassManager.$anonfun$runPassesOnModule$2(PassManager.scala:101)
	at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:183)
	at org.enso.runtime/scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:179)
	at org.enso.runtime/scala.collection.immutable.List.foldLeft(List.scala:79)
	at org.enso.runtime/org.enso.compiler.pass.PassManager.runPassesOnModule(PassManager.scala:92)
	at org.enso.runtime/org.enso.compiler.Compiler.runGlobalTypingPasses(Compiler.scala:806)
	at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$7(Compiler.scala:309)
	at org.enso.runtime/org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$7$adapted(Compiler.scala:294)
	at org.enso.runtime/scala.collection.immutable.List.foreach(List.scala:333)
	at org.enso.runtime/org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:294)
	at org.enso.runtime/org.enso.compiler.Compiler.go$1(Compiler.scala:225)
	at org.enso.runtime/org.enso.compiler.Compiler.runInternal(Compiler.scala:232)
	at org.enso.runtime/org.enso.compiler.Compiler.run(Compiler.scala:127)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.$anonfun$compile$1(EnsureCompiledJob.scala:252)
	at org.enso.runtime/cats.syntax.EitherObjectOps$.catchNonFatal$extension(either.scala:391)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.compile(EnsureCompiledJob.scala:246)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.ensureCompiledModule(EnsureCompiledJob.scala:102)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.$anonfun$ensureCompiledFiles$2(EnsureCompiledJob.scala:82)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:246)
	at org.enso.runtime/scala.collection.immutable.List.map(List.scala:79)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.ensureCompiledFiles(EnsureCompiledJob.scala:82)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.run(EnsureCompiledJob.scala:58)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.run(EnsureCompiledJob.scala:41)
	at org.enso.runtime/org.enso.interpreter.instrument.execution.JobExecutionEngine.$anonfun$runInternal$1(JobExecutionEngine.scala:124)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.access$001(PolyglotThread.java:53)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$1.execute(PolyglotThread.java:106)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.executeImpl(PolyglotThread.java:140)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.execute(PolyglotThread.java:131)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:746)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:602)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:519)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:500)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.run(PolyglotThread.java:102)
[ERROR] [2023-12-01T13:37:59+10:00] [enso.org.enso.interpreter.service.ExecutionService] Error executing org.enso.interpreter.instrument.job.EnsureCompiledJob@66e1fdfa
org.enso.compiler.core.CompilerError: Compiler Internal Error: Empty cache preference metadata local.AOC_2023.Main
	at org.enso.runtime/org.enso.compiler.core.Implicits$AsMetadata.$anonfun$unsafeGetMetadata$1(Implicits.scala:109)
	at org.enso.runtime/scala.Option.getOrElse(Option.scala:201)
	at org.enso.runtime/org.enso.compiler.core.Implicits$AsMetadata.unsafeGetMetadata(Implicits.scala:109)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.$anonfun$getCacheMetadata$2(EnsureCompiledJob.scala:503)
	at org.enso.runtime/scala.Option.map(Option.scala:242)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.$anonfun$getCacheMetadata$1(EnsureCompiledJob.scala:499)
	at org.enso.runtime/scala.Option.flatMap(Option.scala:283)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.getCacheMetadata(EnsureCompiledJob.scala:496)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.$anonfun$setCacheWeights$1(EnsureCompiledJob.scala:474)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.$anonfun$setCacheWeights$1$adapted(EnsureCompiledJob.scala:473)
	at org.enso.runtime/scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at org.enso.runtime/scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at org.enso.runtime/scala.collection.AbstractIterable.foreach(Iterable.scala:933)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.setCacheWeights(EnsureCompiledJob.scala:473)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.run(EnsureCompiledJob.scala:59)
	at org.enso.runtime/org.enso.interpreter.instrument.job.EnsureCompiledJob.run(EnsureCompiledJob.scala:41)
	at org.enso.runtime/org.enso.interpreter.instrument.execution.JobExecutionEngine.$anonfun$runInternal$1(JobExecutionEngine.scala:124)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.access$001(PolyglotThread.java:53)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$1.execute(PolyglotThread.java:106)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.executeImpl(PolyglotThread.java:140)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread$ThreadSpawnRootNode.execute(PolyglotThread.java:131)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:746)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callBoundary(OptimizedCallTarget.java:602)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:586)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:519)
	at org.graalvm.truffle.runtime/com.oracle.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:500)
	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotThread.run(PolyglotThread.java:102)
corrupted size vs. prev_size while consolidating
</code></pre>
</details>

### Enso Version

develop

### Browser or standalone distribution

Chrome

### Browser Version or standalone distribution

119.0.6045.159

### Operating System

Linux

### Operating System Version

Linux 6.6.2

### Hardware you are using

_No response_

--- a/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/desugar/SectionsToBinOp.scala
+++ b/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/desugar/SectionsToBinOp.scala
@@ -81,7 +81,7 @@ case object SectionsToBinOp extends IRPass {
     )
 
     ir.transformExpressions { case sec: Section =>
-      desugarSections(sec, freshNameSupply)
+      desugarSections(sec, freshNameSupply, inlineContext)
     }
   }
 
@@ -97,7 +97,8 @@ case object SectionsToBinOp extends IRPass {
     */
   private def desugarSections(
     section: Section,
-    freshNameSupply: FreshNameSupply
+    freshNameSupply: FreshNameSupply,
+    inlineContext: InlineContext
   ): Expression = {
     section match {
       case Section.Left(arg, op, loc, passData, diagnostics) =>
@@ -145,9 +146,11 @@ case object SectionsToBinOp extends IRPass {
           )
 
         } else {
+          val newArg = arg.mapExpressions(runExpression(_, inlineContext))
+
           Application.Prefix(
             function             = op,
-            arguments            = List(arg),
+            arguments            = List(newArg),
             hasDefaultsSuspended = false,
             location             = loc,
             passData,
@@ -265,9 +268,11 @@ case object SectionsToBinOp extends IRPass {
             loc
           )
         } else {
+          val newArg = arg.mapExpressions(runExpression(_, inlineContext))
+
           val opCall = Application.Prefix(
             function             = op,
-            arguments            = List(leftCallArg, arg),
+            arguments            = List(leftCallArg, newArg),
             hasDefaultsSuspended = false,
             location             = None,
             passData,

--- a/engine/runtime/src/test/java/org/enso/compiler/ExecCompilerTest.java
+++ b/engine/runtime/src/test/java/org/enso/compiler/ExecCompilerTest.java
@@ -64,6 +64,46 @@ public class ExecCompilerTest {
   }
 
   @Test
+  public void testDesugarOperators() throws Exception {
+    var module = ctx.eval("enso", """
+    main =
+      ma ==ums==
+    """);
+    try {
+      var run = module.invokeMember("eval_expression", "main");
+      fail("Unexpected result: " + run);
+    } catch (PolyglotException ex) {
+      assertEquals("Compile error: The name `ma` could not be found.", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testDesugarOperatorsLeftRight() throws Exception {
+    var module = ctx.eval("enso", """
+    main = (+ (2 *))
+    """);
+    try {
+      var run = module.invokeMember("eval_expression", "main");
+      fail("Unexpected result: " + run);
+    } catch (PolyglotException ex) {
+      assertEquals("Method `+` of type Unnamed could not be found.", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testDesugarOperatorsRightLeft() throws Exception {
+    var module = ctx.eval("enso", """
+    main = ((* 2) +)
+    """);
+    try {
+      var run = module.invokeMember("eval_expression", "main");
+      fail("Unexpected result: " + run);
+    } catch (PolyglotException ex) {
+      assertEquals("Method `+` of type Function could not be found.", ex.getMessage());
+    }
+  }
+
+  @Test
   public void testHalfAssignment() throws Exception {
     var module =
         ctx.eval(

